rust/rsmgp-sys/src/value/mod.rs

```diff
@@ -0,0 +1,326 @@
+use std::ffi::{CStr, CString};
+
+use crate::context::*;
+use crate::edge::*;
+use crate::mgp::*;
+use crate::path::*;
+use crate::result::*;
+use crate::vertex::*;
+// Required here, if not present tests linking fails.
+#[double]
+use crate::mgp::ffi;
+use mockall_double::double;
+
+// mgp_value used to return data is non-const, owned by the module code. Function to delete
+// mgp_value is non-const.
+// mgp_value containing data from Memgraph is const.
+//
+// `make` functions return non-const value that has to be deleted.
+// `get_property` functions return non-const copied value that has to be deleted.
+// `mgp_property` holds *const mgp_value.
+//
+// Possible solutions:
+//   * An enum containing *mut and *const can work but the implementation would also contain
+//   duplicated code.
+//   * A generic data type seems complex to implement https://stackoverflow.com/questions/40317860.
+//   * Holding a *const mgp_value + an ownership flag + convert to *mut when delete function has to
+//   be called.
+//   * Hold only *mut mgp_value... as soon as there is *const mgp_value, make a copy (own *mut
+//   mgp_value). The same applies for all other data types, e.g. mgp_edge, mgp_vertex.
+//   mgp_value_make_vertex accepts *mut mgp_vartex (required to return user data), but data from
+//   the graph is all *const T.
+
+/// Useful to hold user created `mgp_value` but also to store values coming from Memgraph, e.g.,
+/// property `mgp_value`s.
+#[derive(Debug)]
+pub struct MgpValue {
+    // It's not wise to create a new Vertex out of the existing value pointer because drop with a
+    // valid pointer will be called multiple times -> double free problem.
+    pub ptr: *mut mgp_value,
+}
+
+impl Drop for MgpValue {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.ptr.is_null() {
+                ffi::mgp_value_destroy(self.ptr);
+            }
+        }
+    }
+}
+
+impl MgpValue {
+    pub fn is_null(&self) -> bool {
+        unsafe { ffi::mgp_value_is_null(self.ptr) != 0 }
+    }
+
+    pub fn is_int(&self) -> bool {
+        unsafe { ffi::mgp_value_is_int(self.ptr) != 0 }
+    }
+
+    pub fn is_bool(&self) -> bool {
+        unsafe { ffi::mgp_value_is_bool(self.ptr) != 0 }
+    }
+
+    pub fn is_string(&self) -> bool {
+        unsafe { ffi::mgp_value_is_string(self.ptr) != 0 }
+    }
+
+    pub fn is_double(&self) -> bool {
+        unsafe { ffi::mgp_value_is_double(self.ptr) != 0 }
+    }
+
+    pub fn is_vertex(&self) -> bool {
+        unsafe { ffi::mgp_value_is_vertex(self.ptr) != 0 }
+    }
+
+    pub fn is_edge(&self) -> bool {
+        unsafe { ffi::mgp_value_is_edge(self.ptr) != 0 }
+    }
+}
+
+pub fn make_null_value(context: &Memgraph) -> MgpResult<MgpValue> {
```

Comment:
This should happen under the hood when calling `MgpResult {value: some_int_value, context: context}`, or however Rust handles object construction

Reply:
`MgpResult` is not a struct, it's a custom type based on [std::result::Result](https://doc.rust-lang.org/std/result/) returned from each call to handle both, the result and any error (take a look under `rsmgp-sys/src/result/mod.rs`). But, valid point, I think I'll move all these methods into the `context::Memgraph`. The final code should look like:
```
memgraph.make_null_value()
```